pkg/azuredisk/nodeserver.go

```diff
@@ -380,19 +380,54 @@ func (d *Driver) NodeExpandVolume(ctx context.Context, req *csi.NodeExpandVolume
 	volSizeBytes := int64(capacityBytes)
 	requestGiB := volumehelper.RoundUpGiB(volSizeBytes)
 
-	args := []string{"-o", "source", "--noheadings", "--mountpoint", req.GetVolumePath()}
-	output, err := d.mounter.Exec.Command("findmnt", args...).Output()
+	volumePath := req.GetVolumePath()
+	if len(volumePath) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "volume path must be provided")
+	}
+
+	// Hopefully devicePath was passed from req
+	devicePath := req.GetStagingTargetPath()
+	if devicePath == "" {
+		args := []string{"-o", "source", "--noheadings", "--mountpoint", volumePath}
+		output, err := d.mounter.Exec.Command("findmnt", args...).Output()
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not determine device path: %v", err)
+		}
+
+		devicePath := strings.TrimSpace(string(output))
+		if len(devicePath) == 0 {
+			return nil, status.Errorf(codes.Internal, "Could not get valid device for mount path: %q", volumePath)
+		}
+	}
+
+	isBlock, err := hostutil.NewHostUtil().PathIsDevice(volumePath)
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Could not determine device path: %v", err)
+		return nil, status.Errorf(codes.NotFound, "failed to determine whether %s is block device: %v", volumePath, err)
+	}
+	if isBlock {
+		// Noop for Block NodeExpandVolume
+		klog.V(4).Infof("NodeExpandVolume succeeded on %v to %s, path check is block so this is a no-op", volumeID, volumePath)
+		return &csi.NodeExpandVolumeResponse{}, nil
 	}
 
-	devicePath := strings.TrimSpace(string(output))
-	if len(devicePath) == 0 {
-		return nil, status.Errorf(codes.Internal, "Could not get valid device for mount path: %q", req.GetVolumePath())
+	volumeCapability := req.GetVolumeCapability()
+	if volumeCapability != nil {
+		// VolumeCapability is optional, if specified, validate it
+		caps := []*csi.VolumeCapability{volumeCapability}
+		if !isValidVolumeCapabilities(caps) {
+			return nil, status.Error(codes.InvalidArgument, "VolumeCapability is invalid.")
+		}
+
+		if blk := volumeCapability.GetBlock(); blk != nil {
```

Comment:
Do we need both "PathIsDevice" and volume capability checks?

Reply:
Yeah I am also wondering about this. That's why I had this log below. So if a path is a not a block but the capability says it is, then we should be alerted here. I can remove it if you feel like this is unnecessary